Remove finish span on close

carlosalberto opened the pull request:
Addresses #291 

`SpanBuilder.startActive(Span, boolean)` and `ScopeManager.activate(Span, bool)` have been kept, but as deprecated methods.

Updated the `opentracing-testbed` code to use this change, and added a new, small, `error-reporting` case.

cc @felixbarny @adriancole @yurishkuro @tedsuo @sjoerdtalsma

felixbarny commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -166,16 +166,40 @@
+         * The span will not be automatically finished when {@link Scope#close()} is called.
+         *
+         * <p>
+         * The returned {@link Scope} supports try-with-resources, but is usually
```
Why is it not usually used with try-with-resources?

carlosalberto replied:
That's related to Yuri's comment about removing it - essentially, that, as opposed to using `ScopeManager.activate()`, here we **need** an usual `try` block to do the cleanup (that is, having a reference to the `Span` to finish it after any error has occurred).

```java
try (Scope scope = tracer.buildSpan("foo").startActive()) {
} catch (Exception e) {
  // scope and scope.span() are not reachable here nor in the finally section :/
}
```

adriancole replied:
If this method isn't deleted, I think it would be cleaner if done similar to census and brave's "scoped span" Instead of messing with scope, mess with span.finish(). In other words, return a Span, and do mention it is a programming error to not call span.finish on the same thread. Another way is to leave it out..

Here's the javadoc from brave's ScopedSpan:

```java
  /**
   * Closes the {@link CurrentTraceContext#newScope(TraceContext) scope} associated with this span,
   * then reports the span complete, assigning the most precise duration possible.
   */
  public abstract void finish();
```

felixbarny commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -166,16 +166,40 @@
+         *         scope = tracer.buildSpan("...").startActive();
          *         // (Do work)
          *         scope.span().setTag( ... );  // etc, etc
+         *     } finally {
```
add catch clause

carlosalberto replied:
Will add it 👍

felixbarny commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -190,16 +214,15 @@
          * @see Scope
          */
-        Scope startActive(boolean finishSpanOnClose);
+        Scope startActive();
```
Could you elaborate?

carlosalberto replied:
Oh, that crossed my mind too. So I think that, if it stays, we would need to be clear about the implications of its usage - that is needs to be used with a standard `try` block, i.e.

```java
Scope scope = tracer.buildSpan("foo").startActive();
try {
} catch (Exception e) {
  Tags.ERROR.set(scope.span(), true);
} finally {
  scope.span().finish();
  scope.close();
}
```

(That's why I used such usage in its updated inline documentation ;) )

I'd be up for keeping it (with the proper documentation), but definitely understand it's not a perfect thing, and I'm open to have it removed if deemed so (in such case I will update the PR).

adriancole replied:
This is a strange method, please remove it.

carlosalberto replied:
Seems we have found agreement on this point, so I will update this PR in the next days to have it removed ;)

adriancole commented on opentracing-api/src/main/java/io/opentracing/ScopeManager.java:
```diff
@@ -26,15 +26,28 @@
 public interface ScopeManager {
 
     /**
+     * @deprecated use {@link #activate} instead.
```
nit move to the bottom and say when it will be removed (ideally next release as would the other deprecated methods added prior.. deprecation pre 1.0 is a strange thing but anyway)

carlosalberto replied:
I don't think there's a date to remove this one (we should discuss this topic, but another time ;) ), but we could say "in the next releases" (and definitely before 1.0 ;) )

adriancole commented on opentracing-api/src/main/java/io/opentracing/ScopeManager.java:
```diff
@@ -26,15 +26,28 @@
     Scope activate(Span span, boolean finishSpanOnClose);
 
+    /**
+     * Make a {@link Span} instance active.
```
add some description of what active means as the word isn't intuitive.

For example "Makes this span visible via commands like (XX) until the corresponding scope is close"

Also, note that it is a programming error to leave the scope unclosed or to attempt to close it on a different thread.

carlosalberto replied:
Thanks, will add such explanation.

adriancole commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -166,16 +166,40 @@
         SpanBuilder withStartTimestamp(long microseconds);
 
         /**
+         * @deprecated use {@link #startActive} instead.
```
same.. move down

carlosalberto commented:
Hey @adriancole 

Thanks for the feedback. So my personal take for your points is:

1. That's not a bad thing, but then people would need to remember to do something like:

```java
Span span = tracer.buildSpan("one").startActive();
try {
} catch (Exception e) {
  ...
  span.finish();
} finally {
   tracer.scopeManager.active().close();
}
```

This being implicit kinda makes me wonder (but definitely thing to consider).

2. A special `Span` is something that would add complexity to the current (relatively simple) API. For example: what happens if this very special `Span` is passed to another thread? I'd leave it as a potential addition in the future.

So the thing about removing `Scope.span()` is probably fine, as long as we have `ScopeManager.activeSpan()` (so we always expose the current active `Span` and the active `Scope` without explicitly linking them together). However, I'd like to hear the opinion of other people here.

Opinions? @felixbarny @yurishkuro @tedsuo

yurishkuro commented:
are we shooting for this?
```java
Span span = tracer.buildSpan("nasty").start();
try (Scope scope = tracer.activate(span)) {
  // do nasty stuff
} catch (Exception e) {
  span.SetTag(...);
  ...
} finally {
  span.finish();
}
```

It makes the usage slightly more verbose than the current api, especially without try-with-resource (only if <=1.6, who cares), but avoids the misuse. `scope.span()` is removed.

felixbarny commented:
I agree with @yurishkuro 
* Remove* `SpanBuilder#startActive`
* Remove* `Scope.span()` (this is only useful in combination with `SpanBuilder#startActive`)
* Remove* `ScopeManager.activate(Span, boolean)`

*deprecate and eventually remove

carlosalberto commented:
Hey,

Was wondering, after the feedback, about this (small variation):

* `SpanBuilder.startActive()` and `Scope.span()` gets removed
* `ScopeManager.activeSpan()` and `ScopeManager.active()` (so we have the two of them available, but *not* linked to each other, thus less chances to have users pass `Scope` objects arounds).
* `Trace.activeSpan()` and `Tracer.activate()` as aliases to `ScopeManager`'s respective methods.

felixbarny commented:
Sounds fine to me

adriancole commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -166,16 +166,40 @@
-         *     try (Scope scope = tracer.buildSpan("...").startActive(true)) {
+         *     Scope scope = null;
+         *     try {
+         *         scope = tracer.buildSpan("...").startActive();
```
please inline this before the try. That removes the null checking later.

Also, as mentioned I think the scope having an accessor to the span is problematic.

carlosalberto commented:
Hey all, 

This PR is updated to reflect the latest requested changes, namely:

* Deprecate `SpanBuilder.startActive(boolean)`, `ScopeManager.activate(Span, boolean)` and `Scope.span()`.
* Removed `SpanBuilder.startActive()` (parameterless overload, which had been added early in this PR).
* Add `ScopeManager.activeSpan()` and leave `ScopeManager.active()` in place (So `ScopeManager.active()` is still reachable, but designed to stay for more advanced use, as well as still exposing `ScopeManager`, in case the user wants to unwrap it himself).
* Add `Tracer.activateSpan()`, as a shorthand for `tracer.scopeManager().activate(span)`.
* Updated the testbed module and inline docs overall.

@yurishkuro @adriancole @felixbarny

tylerbenson commented on opentracing-testbed/src/test/java/io/opentracing/testbed/activate_deactivate/README.md:
```diff
@@ -8,7 +8,8 @@ This example shows the usage of the `Continuation` concept (now part of the `Aut
     public void run() {
-	try (Scope scope = tracer.buildSpan("parent").startActive(false)) {
+	Span span = tracer.buildSpan("parent").start();
+	try (Scope scope = tracer.activateSpan(span)) {
```
I thought the span must be manually finished now.  (There are other cases of this same problem elsewhere in this PR too.)

carlosalberto replied:
Hey Tyler - the *activate deactivate* and *multiple callbacks* examples use the more exoteric `AutoFinishScopeManager`, which was actually ignoring the `finishSpanOnClose` parameter anyway, doing the `Span.finish()` call by itself eventually.

I think we could keep this examples along with this specific `ScopeManager` implementation for now, as it shows a very, very specific case (having to finish a `Span` being active in more than one thread/callback). Observe this implementation has its specific and well-known limitations and, IIRC, was kept for the `opentracing-java-v030` shim layer and also for potential usage along with `Scala` frameworks, given their callback nature.

Another options is to refactor the `AutoFinishScopeManager` code to have a similarity with the new API in all ways (for example, by adding a new method that does the conditional finish in the `Span`, maybe `Scope.Continuation.finishSpan()` or similar, deprecating the current API there). *If* we do this part, I'd suggest leaving it for another PR, as I think it's time to merge the current one (I can propose myself to do that change).

Let me know ;)

tylerbenson replied:
Ah, I didn't notice the distinction in context here.  Carry on.

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/ScopeManager.java:
```diff
@@ -26,26 +26,96 @@
+     *     } catch (Exception e) {
+     *         span.log(...);
+     *     } finally {
+     *         span.finish(); // Optionally finish the Span.
```
`// Optionally finish the Span if the operation it represents is logically completed at this point.`

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/ScopeManager.java:
```diff
@@ -26,26 +26,96 @@
+     * Return the currently active {@link Span}.
+     *
+     * <p>
+     * If there is an {@link Span non-null active span}, it becomes an implicit parent
```
Is it necessary to have this mentioned here? It think this comment only belongs to the span builder methods. Scope Manager as an abstraction is not tied to the builder's behavior.

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/ScopeManager.java:
```diff
@@ -26,26 +26,96 @@
+     *
+     * <p>
+     * Because both {@link #active()} and {@link #activeSpan()} reference the current
+     * the active state, they both will be either null or non-null.
```
s/the active/active/

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -26,11 +26,19 @@
+     * This is a shorthand for {@code Tracer.scopeManager().activate(span)}.
+     *
+     * @return a {@link Scope} instance to control the end of the active period for the {@link Span}. It is a
+     * programming error to neglect to call {@link Scope#close()} on the returned instance.
```
and may lead to memory leaks and the scope may remain in the thread-local stack.

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -40,19 +48,20 @@
-     *   // It's also possible to create Spans manually, bypassing the ScopeManager activation.
-     *   Span http = tracer.buildSpan("HandleHTTPRequest")
+     *   // Note: if there is a `tracer.activeSpan()` instance, it will be used as the target
+     *   // of an implicit CHILD_OF Reference when `start()` is invoked.
```
unless another span reference is explicitly provided to the builder

carlosalberto commented:
@yurishkuro Thanks for the edit notes, appreciated. Please take a look whenever you can ;)

carlosalberto commented:
Hey @felixbarny @adriancole Wondering if you are fine with this PR? If yes, I'd like to have it merged, and probably have a Release Candidate to test things out soon after.

@yurishkuro Docs edits look good to you? ;)

felixbarny commented:
LGTM

tylerbenson commented on opentracing-api/src/main/java/io/opentracing/ScopeManager.java:
```diff
@@ -26,26 +26,92 @@
      *
      * @return the {@link Scope active scope}, or null if none could be found.
      */
     Scope active();
```
I thought we were talking about deprecating this method.

carlosalberto commented:
Thanks @felixbarny ! 

I will proceed to merge this PR tomorrow as if seems to have found a proper trade-off of features, unless there's any complain from anybody ;)

carlosalberto commented:
Merged, thanks everybody for their feedback! Remember that this is not a final change, and it could be changed if/as needed (we will likely be doing a Release Candidate to test this and other changes out) ;)

sjoerdtalsma commented:
@carlosalberto I didn't participate in this PR as my opinion is already voiced by others, but I forgot to thank you for creating this!
So thanks :smile:
